tests/ctst/HOW_TO_WRITE_TESTS.md

```diff
@@ -0,0 +1,121 @@
+# How to write a test that runs in parallel
+
+Writting intergation / behavior tests that run in parallel is not trivial.
+This page lists all the rules to follow when writing a test that runs in
+parallel with other tests.
+
+*This is specific to Zenko, but the rules are applicable for any suite of*
+*tests in parallel.*
+
+**Rule #1**
+
+> Tests must be idempotent.
+
+In parallel, the tests can run in any order. It's not because a test works at a
+given time, that it will work if the test order changes after adding new ones
+or renaming files. You must ensure the scenarios are fully idempotent.
+"Flakies" are mostly due to non-idempotent tests, then bugs. It is rarely due
+to the platform.
+
+**Rule #2**
+
+> Use unique resources names. S3 Bucket, Account and S3 Objects should be
+> unique.
+
+It limits the risks of collision between tests.
+For cross-account scenarios, use unique S3 Policies and S3 Roles.
+
+If data is used, such as lock files, ensure they are named appropriately.
+
+**Rule #3**
+
+> Avoid reconfiguring the environment during test execution.
+
+If a reconfiguration triggers a reconciliation of the resources, running tests
+might be affected. Also, it restarts the services, which leads to missing logs,
+harder debugging, and longer test execution.
+
+Examples:
+
+- Location creation;
+- Bucket Website endpoints.
+
+... must be created before the test start, or in a dedicated set of test.
+
+**Rule #4**
+
+> Do not use `atMostOnePicklePerTag`.
+
+If a set of scenario requires the use of `atMostOnePicklePerTag`, they won't
+be executed in parallel, which is:
+
+- Not realistic for the production environment.
+- Not efficient for the test execution: the duration of tests will suffer.
+- A proof that the test is not following **Rule #1**.
+
+Ensuring this rule may be hard, and idempotent scenarios are not always
+possible. Solutions exist:
+
+- Pre-create the resources required for the tests before they start.
+- Use `Before` / `After` hooks to setup your environment once.
+  Concurrency can be controlled with lock files.
+- Make your assertions more flexible: strict and absolute checks are good
+  for unit or functional testing, but in integration tests, prefer using
+  relative checks.
+- As a last resort, we might have a dedicated test suite.
+
+**Rule #5**
+
+> Avoid scenarios with no value.
+
+It is tempting to write scenarios that test more than what is needed
+at the integration testing layer: please trust your unit and functional
```

Comment:
> integration testing layer

this is not really "integration tests" : as far as semantics go, these tests are more akin to "acceptance tests" and "smoke tests".... which we rely on to validate that components are properly integrated.
Best rephrase to avoid the confusion IMHO.

> please trust your unit and functional

"trust" is probably not what we want : if we really trusted, we wouldn't do any test :-)
```
please use unit and functional tests to fully validate the behavior of each component; and e2e tests only to ensure that the product-level feature is working as expected per the requirements. Making good use of unit and functional tests makes testing all cases, including corner cases or error handling, must easier, leading to higher quality; and improves velocity by having a much faster feedback loop.
```